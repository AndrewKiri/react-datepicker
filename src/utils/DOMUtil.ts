import { IDatePicker } from '../common/@types';
const convertPx = (value: number) => `${value}px`;
/**
 * Getting Div position as far as distance
 * @param node
 * @param direction
 * @param distance
 */
export const getDivPosition = (
  node: HTMLDivElement | null,
  direction: IDatePicker.PickerDirection = IDatePicker.PickerDirection.BOTTOM,
  distance: number = 5
): IDatePicker.Position => {
  if (!node) return { left: '', top: '', bottom: '' };

  let top = 0;
  let left = 0;
  let bottom = 0;

  switch (direction) {
    case IDatePicker.PickerDirection.BOTTOM:
      top = node.offsetTop + node.offsetHeight + distance;
      left = node.offsetLeft;
      break;
    case IDatePicker.PickerDirection.TOP:
      bottom = node.offsetTop - distance;
      left = node.offsetLeft;
      break;
  }
  return {
    top: convertPx(top),
    left: convertPx(left),
    bottom: convertPx(bottom),
  };
};
