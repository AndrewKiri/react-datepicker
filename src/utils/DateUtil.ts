import { chunk, range, repeat } from 'lodash';
import * as moment from 'moment';

export const getDayMatrix = (year?: number, month?: number): string[][] => {
  const updateYear = year || moment().year();
  const updateMonth = month || moment().month();

  const date = moment({ years: updateYear, months: updateMonth });

  const startOfMonth = parseInt(date.startOf('month').format('DD'), 10);
  const endOfMonth = parseInt(date.endOf('month').format('DD'), 10);

  const startDay = date.startOf('month').day();
  const remain = (startDay + endOfMonth) % 7;

  return chunk(
    [
      ...repeat(' ', startDay).split(''),
      ...range(startOfMonth, endOfMonth + 1).map(v => `${v}`),
      ...(7 - remain === 7 ? [] : repeat(' ', 7 - remain).split('')),
    ],
    7
  );
};

export const getMonthMatrix = (locale: string) => {
  const months = moment.localeData(locale).monthsShort();
  return chunk(months, 3);
};

export const getYearMatrix = (year: number) => {
  return chunk(range(year - 4, year + 5).map(v => `${v}`), 3);
};

export const isDayEqual = (day1?: moment.Moment, day2?: moment.Moment) => {
  if (!day1 || !day2) return false;
  return day1.format('YYYYMMDD') === day2.format('YYYYMMDD');
};

export const isDayRange = (date: moment.Moment, start?: moment.Moment, end?: moment.Moment) => {
  if (!start || !end) return false;
  const updateDate = date
    .hour(0)
    .minute(0)
    .second(0);
  const updateStart = start
    .hour(0)
    .minute(0)
    .second(0);
  const updateEnd = end
    .hour(0)
    .minute(0)
    .second(0);

  return date.diff(updateStart, 'days') > 0 && date.diff(updateEnd, 'days') < 0;
};
