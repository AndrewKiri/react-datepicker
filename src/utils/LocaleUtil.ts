import * as dayjs from 'dayjs';
import { range } from 'lodash';
import * as localeData from 'dayjs/plugin/localeData';
import * as localizedFormat from 'dayjs/plugin/localizedFormat';

dayjs.extend(localeData);
dayjs.extend(localizedFormat);

export const getMonthShort = (locale: string) => {
  dayjs.locale(locale);
  return range(0, 12).map(v =>
    dayjs()
      .localeData()
      .monthsShort(dayjs().month(v))
  );
};

export const getWeekDays = (locale: string) => {
  dayjs.locale(locale);
  return range(7).map(v =>
    dayjs()
      .localeData()
      .weekdaysShort(dayjs().day(v))
  );
};

export const getToday = (locale: string) => {
  return dayjs()
    .locale(locale)
    .format('LL');
};
