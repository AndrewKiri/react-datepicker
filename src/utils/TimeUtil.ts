import { lpad } from './StringUtil';
import { IDatePicker } from '../common/@types';

export const getMomentHour = (hour: number, type: string) => {
  let updateHour = hour;
  if (hour === 12) {
    updateHour = type === IDatePicker.TimeType.AM ? 0 : 12;
  } else {
    updateHour = type === IDatePicker.TimeType.AM ? hour : hour + 12;
  }
  return updateHour;
};

export const getNormalHour = (hour: number) => {
  let updateHour = hour;
  if (updateHour === 0) {
    updateHour = 12;
  } else if (hour > 12) {
    updateHour = hour - 12;
  }
  return updateHour;
};

export const getTimeType = (hour: number): IDatePicker.TimeType =>
  hour > 11 ? IDatePicker.TimeType.PM : IDatePicker.TimeType.AM;

export const formatTime = (hour: number, minute: number, type: string) =>
  `${lpad(String(hour), 2)}:${lpad(String(minute), 2)} ${type}`;

export const isValidTime = (time: string) => /^(1[0-2]|0?[1-9]):[0-5][0-9] (AM|PM)$/.test(time);

export const parseTime = (time: string) => {
  const hour = time.split(':')[0];
  const minute = time.split(':')[1].substring(0, 2);
  const type = time.split(' ')[1];
  return {
    type: type === IDatePicker.TimeType.AM ? IDatePicker.TimeType.AM : IDatePicker.TimeType.PM,
    hour: parseInt(hour, 10),
    minute: parseInt(minute, 10),
  };
};
