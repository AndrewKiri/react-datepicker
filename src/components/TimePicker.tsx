import * as React from 'react';
import * as classNames from 'classnames';
import { lpad } from '../utils/StringUtil';
import TimeContainer from './TimeContainer';
import { IDatePicker } from '../common/@types';
import { Omit } from '../utils/TypeUtil';
import { ifExistCall } from '../utils/FunctionUtil';
import { getDivPosition } from '../utils/DOMUtil';
import PickerInput, { Props as InputProps } from './PickerInput';
import Backdrop from './Backdrop';

interface TimePickerProps {
  /** Timepicker change event */
  onChange?: (hour: number, minute: number, type: string) => void;
  /** Timepicker default time icon show or hide */
  showDefaultIcon?: boolean;
  /** Timepicker portal version */
  portal?: boolean;
}

interface State {
  timeShow: boolean;
  inputValue: string;
  position: IDatePicker.Position;
}

const inputFormat = (hour: number, minute: number, type: string) =>
  `${lpad(hour.toString(), 2)}:${lpad(minute.toString(), 2)} ${type}`;

type Props = TimePickerProps & Omit<InputProps, 'onChange'>;
class TimePicker extends React.Component<Props, State> {
  public state = {
    timeShow: false,
    inputValue: inputFormat(1, 0, 'AM'),
    position: {
      left: '',
      top: '',
    },
  };

  public inputRef: React.RefObject<HTMLDivElement>;

  constructor(props: Props) {
    super(props);
    this.inputRef = React.createRef();
  }

  public handleTimeContainer = (e: React.MouseEvent) => {
    const { disabled } = this.props;
    if (disabled) {
      return;
    }
    this.setState({
      timeShow: true,
      position: getDivPosition(this.inputRef.current),
    });
  };

  public handleChange = (hour: number, minute: number, type: string) => {
    const { onChange } = this.props;
    this.setState({
      ...this.state,
      inputValue: inputFormat(hour, minute, type),
    });

    ifExistCall(onChange, hour, minute, type);
  };

  public handleInputChange = (e: React.FormEvent<HTMLInputElement>) => {
    this.setState({
      ...this.state,
      inputValue: e.currentTarget.value,
    });
  };

  public handleInputClear = () => {
    this.setState({
      ...this.state,
      inputValue: '',
    });
  };

  public hideTimeContainer = () => {
    this.setState({
      ...this.state,
      timeShow: false,
    });
  };

  public renderInputcomponent = (): JSX.Element | null => {
    const { inputValue } = this.state;
    return (
      <PickerInput
        {...this.props}
        onChange={this.handleInputChange}
        onClear={this.handleInputClear}
        icon={this.props.showDefaultIcon ? <i className="icon icon-time" /> : undefined}
        value={inputValue}
      />
    );
  };

  public render() {
    const {
      timeShow,
      position: { top, left },
    } = this.state;
    const { portal } = this.props;
    let position;
    if (!portal) {
      position = { top, left };
    }
    return (
      <div className="timepicker">
        <div className="timepicker__input" onClick={this.handleTimeContainer} ref={this.inputRef}>
          {this.renderInputcomponent()}
        </div>
        <div
          className={classNames('timepicker__container', { portal })}
          style={{ ...position, display: timeShow ? 'block' : 'none' }}
        >
          <TimeContainer onChange={this.handleChange} />
        </div>
        <Backdrop show={timeShow} invert={portal} onClick={this.hideTimeContainer} />
      </div>
    );
  }
}

export default TimePicker;
