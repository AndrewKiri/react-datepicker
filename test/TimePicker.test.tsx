import { mount, shallow, ShallowWrapper, ReactWrapper } from 'enzyme';
import { mountInputSimulateChange } from './utils/TestingUtil';
import * as React from 'react';
import * as sinon from 'sinon';
import TimePicker from '../src/components/TimePicker';

describe('<TimePicker/>', () => {
  const pickerShow = (component: ReactWrapper) => {
    component.find('Picker').setState({
      show: true,
    });
  };
  describe('shallow render test', () => {
    let shallowComponent: ShallowWrapper<React.Component>;

    beforeEach(() => {
      shallowComponent = shallow(<TimePicker />);
    });

    it('render with no props', () => {
      expect(shallowComponent).toBeTruthy();
      expect(shallowComponent).toMatchSnapshot();
    });
  });

  describe('mount test', () => {
    let onChange: sinon.SinonSpy;
    let mountComponent: ReactWrapper;
    beforeEach(() => {
      onChange = sinon.spy();
      mountComponent = mount(<TimePicker onChange={onChange} />);
      pickerShow(mountComponent);
    });

    it('should props onChange correctly', () => {
      mountComponent
        .find('.time-input__up button')
        .at(0)
        .simulate('click');
      expect(onChange).toHaveProperty('callCount', 1);
    });

    it('should input change event correctly', () => {
      mountComponent.find('.picker__trigger input').simulate('change', {
        currentTarget: {
          value: 'test',
        },
      });
    });
  });

  describe('input props', () => {
    let mountComponent: ReactWrapper;
    beforeEach(() => {
      mountComponent = mount(<TimePicker readOnly disabled clear showDefaultIcon />);
    });

    it('should props disabled not run handleTimeContainer', () => {
      mountComponent.find('.picker-input__text').simulate('click');
      expect(mountComponent.find('Picker').state('show')).toBeFalsy();
    });

    it('should clear button click inputValue empty', () => {
      mountComponent
        .find('.icon-clear')
        .first()
        .simulate('click');
      expect(mountComponent.state('inputValue')).toEqual('');
    });

    it('should props showDefaultIcon correctly', () => {
      expect(mountComponent.find('.icon-time').first()).toHaveLength(1);
    });
  });

  describe('handleInputBlur', () => {
    let mountComponent: ReactWrapper;
    let timepickerInput: ReactWrapper;

    beforeEach(() => {
      mountComponent = mount(<TimePicker />);
      timepickerInput = mountComponent.find('.picker-input__text');
    });

    it('should user invalid input after blur recover original', () => {
      mountComponent.setState({
        ...mountComponent.state,
      });
      pickerShow(mountComponent);

      mountInputSimulateChange(timepickerInput, '233232');
      timepickerInput.simulate('blur');

      expect(mountComponent.state('inputValue')).toEqual('01:00 AM');
    });

    it('should user valid input after blur set state inputValue', () => {
      mountComponent.setState({
        ...mountComponent.state,
      });
      pickerShow(mountComponent);

      mountInputSimulateChange(timepickerInput, '02:22 AM');
      timepickerInput.simulate('blur');
      expect(mountComponent.state('inputValue')).toEqual('02:22 AM');
    });
  });
});
