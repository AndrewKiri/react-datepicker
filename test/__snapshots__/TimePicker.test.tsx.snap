// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimePicker/> shallow render test render with no props 1`] = `
<Picker
  renderContents={[Function]}
  renderTrigger={[Function]}
/>
`;
